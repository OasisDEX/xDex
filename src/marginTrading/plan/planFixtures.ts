import { BigNumber } from 'bignumber.js'
import { Offer, OfferType } from '../../exchange/orderbook/orderbook'
import { zero } from '../../utils/zero'
import {
  CashAssetCore,
  MarginableAssetCore,
  MTAccount,
  // NonMarginableAssetCore
} from '../state/mtAccount'
import {
  getCashCore,
  getMarginableCore,
  getMTAccount,
  // getNonMarginableCore
} from '../state/mtTestUtils'

export const noCash: CashAssetCore = getCashCore({
  name: 'DAI',
  balance: zero,
  walletBalance: zero,
})

export const cash: CashAssetCore = {
  ...noCash,
  balance: new BigNumber('30000'),
}

export const wethEmpty: MarginableAssetCore = getMarginableCore({
  name: 'WETH',
  referencePrice: new BigNumber('200'),
  minCollRatio: new BigNumber('1.5'),
  safeCollRatio: new BigNumber('2'),
})

export const wethEmptyWithDai: MarginableAssetCore = getMarginableCore({
  name: 'WETH',
  referencePrice: new BigNumber('200'),
  minCollRatio: new BigNumber('1.5'),
  safeCollRatio: new BigNumber('2'),
  dai: new BigNumber(30000),
})

export const weth100: MarginableAssetCore = getMarginableCore({
  name: 'WETH',
  referencePrice: new BigNumber('200'),
  minCollRatio: new BigNumber('1.5'),
  safeCollRatio: new BigNumber('2'),
  balance: new BigNumber('100'),
})

// export const mkrEmpty: NonMarginableAssetCore = getNonMarginableCore({
//   name: 'MKR',
//   referencePrice: new BigNumber('200'),
// });
//
// export const mkr100: NonMarginableAssetCore = getNonMarginableCore({
//   ...mkrEmpty,
//   balance: new BigNumber('100'),
// });

export const dgx: MarginableAssetCore = getMarginableCore({
  name: 'DGX',
  referencePrice: new BigNumber('50'),
  minCollRatio: new BigNumber('1.1'),
  safeCollRatio: new BigNumber('1.2'),
})

export const dgx100 = {
  ...dgx,
  balance: new BigNumber('100'),
}

export const mtaOnlyWeth: MTAccount = getMTAccount({ cash, marginableAssets: [wethEmpty] })

export const sell1: Offer = {
  offerId: new BigNumber('1'),
  baseAmount: new BigNumber('100'),
  baseToken: 'WETH',
  quoteAmount: new BigNumber('20000'),
  quoteToken: 'DAI',
  price: new BigNumber('200'),
  ownerId: 'koko',
  timestamp: new Date(),
  type: OfferType.sell,
}

export const sell2: Offer = {
  ...sell1,
  baseAmount: new BigNumber('100'),
  quoteAmount: new BigNumber('20100'),
  price: new BigNumber('201'),
}

export const sell3: Offer = {
  ...sell1,
  baseAmount: new BigNumber('100'),
  quoteAmount: new BigNumber('20200'),
  price: new BigNumber('202'),
}

export const sellOffers = [sell1, sell2, sell3]
