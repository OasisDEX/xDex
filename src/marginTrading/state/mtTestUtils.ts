import { BigNumber } from 'bignumber.js';

import { AssetKind } from '../../blockchain/config';
import {
  CashAssetCore,
  MarginableAssetCore,
  MTAccount, MTAccountSetup,
  MTAccountState,
  NonMarginableAssetCore
} from './mtAccount';
import { calculateMTAccount } from './mtCalculate';

export function getCashCore(props?: Partial<CashAssetCore>): CashAssetCore {
  return {
    name: 'DAI',
    balance: new BigNumber(0),
    walletBalance: new BigNumber(0),
    marginBalance: new BigNumber(0),
    allowance: true,
    assetKind: AssetKind.cash,
    ...props,
  } as CashAssetCore;
}

export function getMarginableCore(props?: Partial<MarginableAssetCore>): MarginableAssetCore {
  return {
    name: 'W-ETH',
    balance: new BigNumber(0),
    urnBalance: new BigNumber(0),
    walletBalance: new BigNumber(0),
    marginBalance: new BigNumber(0),
    allowance: true,
    assetKind: AssetKind.marginable,
    debt: new BigNumber(0),
    referencePrice: new BigNumber(0),
    minCollRatio: new BigNumber(0),
    safeCollRatio: new BigNumber(0),
    rate: new BigNumber(0),
    history: [],
    ...props
  };
}

export function getNonMarginableCore(
  props?: Partial<NonMarginableAssetCore>
): NonMarginableAssetCore {
  return {
    name: 'MKR',
    balance: new BigNumber(0),
    walletBalance: new BigNumber(0),
    marginBalance: new BigNumber(0),
    allowance: true,
    assetKind: AssetKind.nonMarginable,
    referencePrice: new BigNumber(0),
    ...props
  };
}

export function getNotSetupMTAccount(): MTAccount {
  return {
    state: MTAccountState.notSetup,
  };
}

export function getMTAccount(props: {
  cash?: Partial<CashAssetCore>;
  marginableAssets?: Array<Partial<MarginableAssetCore>>;
  nonMarginableAssets?: Array<Partial<NonMarginableAssetCore>>;
} = {}): MTAccountSetup {
  return calculateMTAccount(
    undefined,
    props.cash ? getCashCore(props.cash) : getCashCore(),
    props.marginableAssets ? props.marginableAssets.map(getMarginableCore) : [],
    props.nonMarginableAssets ? props.nonMarginableAssets.map(getNonMarginableCore) : []
  );
}
