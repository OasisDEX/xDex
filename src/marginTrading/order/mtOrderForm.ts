import { BigNumber } from 'bignumber.js';
// import { curry } from 'lodash';
import { merge, Observable, Subject } from 'rxjs';
// import { tap } from 'rxjs/internal/operators';
import {
  map,
  scan
} from 'rxjs/operators';
import { DustLimits } from '../../balances-nomt/balances';
import { Calls$ } from '../../blockchain/calls/calls';
import { AssetKind, tokens } from '../../blockchain/config';
import { OfferType, Orderbook } from '../../exchange/orderbook/orderbook';
import { TradingPair } from '../../exchange/tradingPair/tradingPair';
import { combineAndMerge } from '../../utils/combineAndMerge';
import {
  AmountFieldChange, DustLimitChange, EtherPriceUSDChange, FormChangeKind,
  FormResetChange, GasEstimationStatus, GasPriceChange, HasGasEstimation, KindChange,
  MTAccountChange, OrderbookChange, PriceFieldChange,
  SetMaxChange, toDustLimitChange$, toEtherPriceUSDChange,
  toGasPriceChange,
  toMTAccountChange, toOrderbookChange$
} from '../../utils/form';
import { description, impossible, Impossible, isImpossible } from '../../utils/impossible';
import { AllocationRequestPilot, DebtDelta } from '../allocate/allocate';
import { prepareBuyAllocationRequest } from '../plan/planBuy';
import { prepareSellAllocationRequest } from '../plan/planSell';
import { getTotal } from '../plan/planUtils';
import {
  findAsset, findMarginableAsset, findNonMarginableAsset,
  MarginableAsset,
  MTAccount,
  MTAccountState,
  NonMarginableAsset,
  Operation
} from '../state/mtAccount';
import {
  realPurchasingPowerMarginable,
  realPurchasingPowerNonMarginable
} from '../state/mtCalculate';

export enum MessageKind {
  insufficientAmount = 'insufficientAmount',
  incredibleAmount = 'incredibleAmount',
  dustAmount = 'dustAmount',
  dustTotal = 'dustTotal',
  impossibleToPlan = 'impossibleToPlan',
  impossibleCalculateTotal = 'impossibleCalculateTotal',
}

export type Message = {
  kind: MessageKind.dustTotal;
  field: string;
  priority: number;
} |  {
  kind: MessageKind.insufficientAmount | MessageKind.incredibleAmount;
  field: string;
  priority: number;
  token: string;
} | {
  kind: MessageKind.dustAmount ;
  field: string;
  priority: number;
  token: string;
  amount: BigNumber;
} | {
  kind: MessageKind.impossibleToPlan | MessageKind.impossibleCalculateTotal;
  field?: string;
  priority: number;
  message: string
};

export enum FormStage {
  editing = 'editing',
  readyToAllocate = 'readyToAllocate',
  waitingForAllocation = 'waitingForAllocation'
}

export interface MTFormState extends HasGasEstimation {
  baseToken: string;
  quoteToken: string;
  kind: OfferType;
  stage: FormStage;
  amount?: BigNumber;
  price?: BigNumber;
  total?: BigNumber;
  messages: Message[];
  orderbook?: Orderbook;
  dustLimitQuote?: BigNumber;
  dustLimitBase?: BigNumber;
  mta?: MTAccount;
  realPurchasingPower?: BigNumber;
  allocationRequest?: AllocationRequestPilot | Impossible;
  debtDeltas?: DebtDelta[];
  plan?: Operation[] | Impossible;
  change: (change: ManualChange) => void;
}

export type ManualChange =
  PriceFieldChange |
  AmountFieldChange |
  SetMaxChange |
  KindChange;

export interface FormStageChange {
  kind: FormChangeKind.formStageChange;
  stage: FormStage;
}

export type StageChange =
  FormResetChange |
  FormStageChange;

export type EnvironmentChange =
  GasPriceChange |
  EtherPriceUSDChange |
  OrderbookChange |
  DustLimitChange |
  MTAccountChange;

export type MTFormChange = ManualChange | StageChange | EnvironmentChange;

function applyChange(state: MTFormState, change: MTFormChange): MTFormState {
  switch (change.kind) {
    case FormChangeKind.amountFieldChange:
      return {
        ...state,
        amount: change.value,
        ...change.value && state.price
          ? { total: change.value.multipliedBy(state.price) }
          : {},
        gasEstimationStatus: GasEstimationStatus.unset
      };
    case FormChangeKind.priceFieldChange:
      return {
        ...state,
        price: change.value,
        ...change.value && state.amount
          ? { total: change.value.multipliedBy(state.amount) }
          : {},
        gasEstimationStatus: GasEstimationStatus.unset
      };
    case FormChangeKind.setMaxChange:
      const baseAsset = findAsset(state.baseToken, state.mta);
      switch (state.kind) {
        case OfferType.sell:
          const baseBalance = (baseAsset && baseAsset.balance) || new BigNumber(0);
          if (state.price) {
            return {
              ...state,
              amount: baseBalance,
              total: baseBalance.times(state.price),
              gasEstimationStatus: GasEstimationStatus.unset
            };
          }
          return {
            ...state,
            amount: baseBalance,
            gasEstimationStatus: GasEstimationStatus.unset
          };
        case OfferType.buy:
          const basePurchasingPower = (baseAsset
            && (baseAsset as MarginableAsset).purchasingPower) || new BigNumber(0);
          if (state.price) {
            return {
              ...state,
              amount: basePurchasingPower.dividedBy(state.price),
              total: basePurchasingPower,
              gasEstimationStatus: GasEstimationStatus.unset
            };
          }
      }
      return {
        ...state,
      };
    case FormChangeKind.kindChange:
      return {
        ...state,
        kind: change.newKind,
        gasEstimationStatus: GasEstimationStatus.unset
      };
    case FormChangeKind.formResetChange:
      return {
        ...state,
        stage: FormStage.editing,
        price: undefined,
        amount: undefined,
        total: undefined
      };
    case FormChangeKind.formStageChange:
      return { ...state, stage: change.stage };
    case FormChangeKind.gasPriceChange:
      return { ...state, gasPrice: change.value };
    case FormChangeKind.etherPriceUSDChange:
      return { ...state,
        etherPriceUsd: change.value,
        gasEstimationStatus: GasEstimationStatus.unset };
    case FormChangeKind.marginTradingAccountChange:
      return { ...state,
        mta: change.mta,
        gasEstimationStatus: GasEstimationStatus.unset };
    case FormChangeKind.dustLimitChange:
      return { ...state,
        dustLimitBase: change.dustLimitBase,
        dustLimitQuote: change.dustLimitQuote
      };
    case FormChangeKind.orderbookChange:
      return { ...state,
        orderbook: change.orderbook };
  }
}

function validate(state: MTFormState): MTFormState {

  if (state.stage !== FormStage.editing) {
    return state;
  }
  const messages: Message[] = [...state.messages];
  const baseAsset = findAsset(state.baseToken, state.mta) as MarginableAsset | NonMarginableAsset;
  const quoteAsset = findAsset(state.quoteToken, state.mta);
  if (
    state.price && state.amount && state.total &&
    baseAsset && quoteAsset && state.realPurchasingPower
  ) {
    const [spendAmount, spendAssetAvailBalance, spendAssetName, spendField, spendDustLimit,
        receiveAmount, receiveAssetName, receiveField] =
      state.kind === OfferType.sell ?
      [state.amount, baseAsset.balance, baseAsset.name, 'amount', state.dustLimitBase,
        state.total, quoteAsset.name, 'total'] :
      [state.total, state.realPurchasingPower, quoteAsset.name, 'total', state.dustLimitQuote,
        state.amount, baseAsset.name, 'amount'];

    if (spendAssetAvailBalance.lt(spendAmount)) {
      messages.push({
        kind: MessageKind.insufficientAmount,
        field: spendField,
        priority: 1,
        token: spendAssetName,
      });
    }
    if ((spendDustLimit || new BigNumber(0)).gt(spendAmount)) {
      messages.push({
        kind: MessageKind.dustAmount,
        field: spendField,
        priority: 2,
        token: spendAssetName,
        amount: spendDustLimit || new BigNumber(0),
      });
    }
    if (new BigNumber(tokens[spendAssetName].maxSell).lt(spendAmount)) {
      messages.push({
        kind: MessageKind.incredibleAmount,
        field: spendField,
        priority: 2,
        token: spendAssetName,
      });
    }
    if (new BigNumber(tokens[receiveAssetName].maxSell).lt(receiveAmount)) {
      messages.push({
        kind: MessageKind.incredibleAmount,
        field: receiveField,
        priority: 1,
        token: receiveAssetName,
      });
    }
    if (state.total.lte(new BigNumber(0))) {
      messages.push({
        kind: MessageKind.dustTotal,
        field: 'total',
        priority: 1,
      });
    }
  }
  return {
    ...state,
    messages,
    gasEstimationStatus: GasEstimationStatus.unset
  };
}

function doAllocationRequest(state: MTFormState): MTFormState {

  if (
    !(state.mta &&
    state.mta.state !== MTAccountState.notSetup &&
    state.orderbook &&
    state.amount &&
    state.price &&
    state.realPurchasingPower)
  ) {
    return {
      ...state,
      allocationRequest: impossible('state not ready'),
    };
  }

  const prepareAllocationRequest = state.kind === OfferType.buy ?
    prepareBuyAllocationRequest :
    prepareSellAllocationRequest;

  const allocationRequest = prepareAllocationRequest(
    state.mta,
    state.kind === OfferType.buy ? state.orderbook.sell : state.orderbook.buy,
    state.baseToken,
    state.amount,
    state.price,
    state.realPurchasingPower
  );

  const messages: Message[] =
      isImpossible(allocationRequest) ?
      [...state.messages,
        {
          kind: MessageKind.impossibleToPlan,
          message: description(allocationRequest),
          priority: 1,
          field: 'total'
        }
      ] :
      state.messages;

  return {
    ...state,
    allocationRequest,
    messages,
  };
}

function checkIfIsReadyToAllocate(state: MTFormState): MTFormState {

  if (state.stage !== FormStage.editing) {
    return state;
  }

  if (
    // state.gasEstimationStatus === GasEstimationStatus.calculated &&
    state.amount && state.price &&
    state.messages.length === 0 &&
    !isImpossible(state.allocationRequest)
  ) {
    return { ...state, stage: FormStage.readyToAllocate };
  }

  return { ...state, stage: FormStage.editing };
}

function addPurchasingPower(state: MTFormState) {

  const baseAsset =
    findMarginableAsset(state.baseToken, state.mta) ||
    findNonMarginableAsset(state.baseToken, state.mta);

  if (!state.mta
    || state.mta.state !== MTAccountState.setup
    || !state.orderbook
    || !baseAsset
  ) {
    return state;
  }

  return {
    ...state,
    realPurchasingPower: baseAsset.assetKind === AssetKind.marginable ?
      realPurchasingPowerMarginable(
        baseAsset,
        state.mta.cash.balance.plus(findMarginableAsset(state.baseToken, state.mta)!.availableDebt),
        state.orderbook.sell)
      :
      realPurchasingPowerNonMarginable(
        state.mta.cash.balance,
        state.orderbook.sell
      )
  };
}

function addTotal(state: MTFormState) {
  if (!state.amount || !state.orderbook) {
    return state;
  }

  const orderbookTotal = getTotal(
    state.amount,
    state.kind === OfferType.buy ?
      state.orderbook.sell :
      state.orderbook.buy);

  if (isImpossible(orderbookTotal)) {
    const messages: Message[] =
      [
        ...state.messages,
        {
          kind: MessageKind.impossibleCalculateTotal,
          message: description(orderbookTotal),
          priority: 1,
          field: 'amount'
        }
      ];

    return {
      ...state,
      messages,
      total: undefined,
      price: undefined,
    };
  }

  const orderbookPrice = orderbookTotal.div(state.amount);

  return {
    ...state,
    total: orderbookTotal,
    price: orderbookPrice
  };
}

export function createMTOrderForm$(
  tradingPair: TradingPair,
  gasPrice$: Observable<BigNumber>,
  etherPriceUSD$: Observable<BigNumber>,
  orderbook$: Observable<Orderbook>,
  mta$: Observable<MTAccount>,
  _calls$: Calls$,
  dustLimits$: Observable<DustLimits>,
): Observable<MTFormState> {

  const manualChange$ = new Subject<ManualChange>();

  const environmentChange$ = combineAndMerge(
    toGasPriceChange(gasPrice$),
    toEtherPriceUSDChange(etherPriceUSD$),
    toOrderbookChange$(orderbook$),
    toDustLimitChange$(dustLimits$, tradingPair.base, tradingPair.quote),
    toMTAccountChange(mta$)
  );

  const initialState: MTFormState = {
    kind: OfferType.buy,
    baseToken: tradingPair.base,
    quoteToken: tradingPair.quote,
    gasEstimationStatus: GasEstimationStatus.unset,
    messages: [],
    stage: FormStage.editing,
    change: manualChange$.next.bind(manualChange$),
  };

  return merge(
    manualChange$,
    environmentChange$,
  ).pipe(
    scan(applyChange, initialState),
    map(addPurchasingPower),
    map(addTotal),
    map(validate),
    map(doAllocationRequest),
    map(checkIfIsReadyToAllocate),
  );
}
