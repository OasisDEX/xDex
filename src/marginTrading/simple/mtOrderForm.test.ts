import { BigNumber } from 'bignumber.js';
import { omit } from 'lodash';
import { of } from 'rxjs';
import { Calls$, ReadCalls$ } from '../../blockchain/calls/calls';
import { setupFakeWeb3ForTesting } from '../../blockchain/web3';
import { createFakeOrderbook, emptyOrderBook } from '../../exchange/depthChart/fakeOrderBook';
import { OfferType } from '../../exchange/orderbook/orderbook';
import { FormChangeKind, GasEstimationStatus } from '../../utils/form';
import { unpack } from '../../utils/testHelpers';
import { noCash, wethEmpty } from '../plan/planFixtures';
import { CashAssetCore, MTAccount } from '../state/mtAccount';
import { getMTAccount } from '../state/mtTestUtils';
import { createMTSimpleOrderForm$, MessageKind, MTSimpleOrderFormParams } from './mtOrderForm';
setupFakeWeb3ForTesting();

function snapshotify(object: any): any {
  return omit(object, ['change', 'timestamp']);
}

const tradingPair = { base: 'WETH', quote: 'DAI' };

const defaultCalls = {
  mtBuyEstimateGas: () => of(20),
  mtSellEstimateGas: () => of(30),
} as any;

const defaultReadCalls = {
} as any;

const defaultUser = '0x1234';

const mtaEmpty: MTAccount = getMTAccount();

const defParams = {
  gasPrice$: of(new BigNumber(0.01)),
  etherPriceUsd$: of(new BigNumber(1)),
  orderbook$: of(emptyOrderBook),
  calls$: of(defaultCalls) as Calls$,
  readCalls$: of(defaultReadCalls) as ReadCalls$,
  dustLimits$: of({ DAI: new BigNumber(0.1), WETH: new BigNumber(0.1) }),
  account$: of(defaultUser),
};

const controllerWithFakeOrderBook = (
  buys: any = [], sells: any = [], mta: MTAccount = mtaEmpty
) => {
  const orderbook = createFakeOrderbook(buys, sells);
  orderbook.buy.forEach((v, i) => v.offerId = new BigNumber(i + 1));
  orderbook.sell.forEach((v, i) => v.offerId = new BigNumber(i + 1));
  return createMTSimpleOrderForm$(
    {
      ...defParams,
      orderbook$: of(orderbook),
      mta$: of(mta),
    } as MTSimpleOrderFormParams,
    tradingPair
  );
};

test('initial state', done => {
  // const controller = createMTSimpleOrderForm$(defParams, tradingPair);
  const sells = [
    { price: 1.3, amount: 3 }, // 1
    { price: 2, amount: 3 }, // 2
    { price: 4, amount: 3 }, // 3
    { price: 5, amount: 4 }, // 4
  ];
  const controller = controllerWithFakeOrderBook([], sells);
  controller.subscribe(state => {
    expect(snapshotify(state)).toMatchSnapshot();
    done();
  });
});

test('set price and amount', () => {
  const sells = [
    { price: 1.3, amount: 3 }, // 1
    { price: 2, amount: 3 }, // 2
    { price: 4, amount: 3 }, // 3
    { price: 5, amount: 4 }, // 4
  ];
  const controller = controllerWithFakeOrderBook([], sells);
  const { change } = unpack(controller);

  change({ kind: FormChangeKind.amountFieldChange, value: new BigNumber(1) });

  expect(unpack(controller).amount).toEqual(new BigNumber(1));
  expect(unpack(controller).price).toEqual(new BigNumber(1.3));
  expect(unpack(controller).total).toEqual(new BigNumber(1.3));
  expect(unpack(controller).gasEstimationStatus).toEqual(GasEstimationStatus.unset);
  expect(snapshotify(unpack(controller))).toMatchSnapshot();
});

test('calculate position in order book for buy', () => {
  const buys = [
    { price: 5, amount: 3 }, // 1
    { price: 4, amount: 3 }, // 2
    { price: 2.5, amount: 3 }, // 3
    { price: 2, amount: 4 }, // 4
    { price: 1, amount: 4 }, // 5
  ];
  const  controller = controllerWithFakeOrderBook(buys);
  expect(unpack(controller).amount).toBeUndefined();
  expect(unpack(controller).total).toBeUndefined();

  // change({ kind: FormChangeKind.amountFieldChange, value: new BigNumber(2) });
  // expect(unpack(controller).position).toBeUndefined();
});

test('calculate undefined position in empty order book for buy', () => {
  const controller = controllerWithFakeOrderBook();
  const { change } = unpack(controller);

  change({ kind: FormChangeKind.amountFieldChange, value: new BigNumber(2) });
  expect(unpack(controller).price).toBeUndefined();
  expect(unpack(controller).total).toBeUndefined();
});

test('calculate position in orderbook for sell', () => {
  const sells = [
    { price: 1.3, amount: 3 }, // 1
    { price: 2, amount: 3 }, // 2
    { price: 4, amount: 3 }, // 3
    { price: 5, amount: 4 }, // 4
  ];

  const controller = controllerWithFakeOrderBook(sells);
  const { change } = unpack(controller);

  change({ kind: FormChangeKind.kindChange, newKind: OfferType.sell });
  expect(unpack(controller).kind).toEqual(OfferType.sell);
  expect(unpack(controller).amount).toBeUndefined();

  change({ kind: FormChangeKind.amountFieldChange, value: new BigNumber(3.5) });
  expect(unpack(controller).price).toEqual(new BigNumber(1.4));
  expect(unpack(controller).total).toEqual(new BigNumber(4.9));
});

test('buy with leverage - match exactly one order', () => {
  const cash: CashAssetCore = {
    ...noCash,
  };
  const weth = {
    ...wethEmpty,
    referencePrice: new BigNumber(1),
    balance: new BigNumber(0),
    debt: new BigNumber(0),
    dai: new BigNumber(1)
  };

  const mta: MTAccount = getMTAccount({ cash, marginableAssets: [weth] });

  const sells = [
    { price: 1, amount: 1 }, // 1
    { price: 1, amount: 1 }, // 1
    { price: 1.2, amount: 1 }, // 1
  ];

  const controller = controllerWithFakeOrderBook([], sells, mta);
  const { change } = unpack(controller);

  expect(unpack(controller).realPurchasingPower).toEqual(new BigNumber(1.984375));
  change({ kind: FormChangeKind.amountFieldChange, value: new BigNumber(1) });

  expect(unpack(controller).readyToProceed).toEqual(true);
  expect(unpack(controller).realPurchasingPowerPost).toEqual(new BigNumber(0.984375));
});

test('buy with leverage - match more than one order', () => {
  const weth = {
    ...wethEmpty,
    referencePrice: new BigNumber(1),
    balance: new BigNumber(2),
    debt: new BigNumber(0)
  };

  const mta: MTAccount = getMTAccount({ marginableAssets: [weth] });

  const sells = [
    { price: 1, amount: 1 }, // 1
    { price: 1.1, amount: 1 }, // 1
    { price: 1.2, amount: 1 }, // 1

  ];

  const controller = controllerWithFakeOrderBook([], sells, mta);
  const { change } = unpack(controller);

  change({ kind: FormChangeKind.amountFieldChange, value: new BigNumber(1.5) });

  expect(unpack(controller).readyToProceed).toEqual(true);
  // expect(unpack(controller).realPurchasingPowerPost).toEqual(new BigNumber(0.484375));
});

test('buy with leverage - purchasing power too low', () => {
  const weth = {
    ...wethEmpty,
    referencePrice: new BigNumber('1'),
    balance: new BigNumber('2'),
    debt: new BigNumber('0')
  };
  const mta: MTAccount = getMTAccount({ marginableAssets: [weth] });
  const sells = [
    { price: 1, amount: 10 }
  ];

  const controller = controllerWithFakeOrderBook([], sells, mta);
  const { change } = unpack(controller);

  expect(unpack(controller).realPurchasingPower).toEqual(new BigNumber(1.984375));
  change({ kind: FormChangeKind.amountFieldChange, value: new BigNumber(3) });
  expect(unpack(controller).messages[0].kind).toEqual(MessageKind.impossibleToPlan);
  expect(unpack(controller).messages[0].message).toEqual('purchasing power too low');
});

test('buy with leverage - orderbook too shallow', () => {
  const weth = {
    ...wethEmpty,
    referencePrice: new BigNumber('1'),
    balance: new BigNumber('20'),
    debt: new BigNumber('0')
  };
  const mta: MTAccount = getMTAccount({ marginableAssets: [weth] });
  const sells = [
    { price: 1, amount: 10 }
  ];

  const controller = controllerWithFakeOrderBook([], sells, mta);
  const { change } = unpack(controller);

  change({ kind: FormChangeKind.amountFieldChange, value: new BigNumber(15) });
  expect(unpack(controller).messages[0].kind).toEqual(MessageKind.impossibleCalculateTotal);
  expect(unpack(controller).messages[0].message).toEqual('orderbook too shallow');
});

test('buy with leverage - collateral and cash', () => {
  const weth = {
    ...wethEmpty,
    referencePrice: new BigNumber(1),
    balance: new BigNumber(5),
    debt: new BigNumber(0),
    dai: new BigNumber(5),
  };
  const mta: MTAccount = getMTAccount({ marginableAssets: [weth] });
  const sells = [
    { price: 1, amount: 20 }
  ];

  const controller = controllerWithFakeOrderBook([], sells, mta);
  const { change } = unpack(controller);

  expect(unpack(controller).leverage).toEqual(new BigNumber(1));
  change({ kind: FormChangeKind.amountFieldChange, value: new BigNumber(14.98) });
  expect(unpack(controller).leveragePost).toEqual(new BigNumber(1.998));
});

test('buy with leverage - cash only', () => {
  const weth = {
    ...wethEmpty,
    referencePrice: new BigNumber(1),
    balance: new BigNumber(0),
    debt: new BigNumber(0),
    dai: new BigNumber(10),
  };
  const mta: MTAccount = getMTAccount({ marginableAssets: [weth] });
  const sells = [
    { price: 1, amount: 20 }
  ];

  const controller = controllerWithFakeOrderBook([], sells, mta);
  const { change } = unpack(controller);

  change({ kind: FormChangeKind.amountFieldChange, value: new BigNumber(15) });
  expect(unpack(controller).leveragePost).toEqual(new BigNumber(1.5));
});
