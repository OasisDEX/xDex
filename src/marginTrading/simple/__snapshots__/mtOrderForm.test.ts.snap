// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`initial state 1`] = `
Object {
  "account": "0x1234",
  "baseToken": "WETH",
  "collRatio": undefined,
  "dustLimitBase": "0.1",
  "dustLimitQuote": "0.1",
  "etherPriceUsd": "1",
  "gasEstimationStatus": "unset",
  "gasPrice": "0.01",
  "isSafeCollRatio": true,
  "kind": "buy",
  "leverage": undefined,
  "liquidationPrice": undefined,
  "maxAmount": "0",
  "maxTotal": "0",
  "messages": Array [],
  "mta": Object {
    "daiAllowance": false,
    "marginableAssets": Array [],
    "proxy": Object {
      "options": Object {
        "address": "",
      },
    },
    "state": "setup",
    "totalAssetValue": "0",
    "totalAvailableDebt": "0",
    "totalDebt": "0",
  },
  "orderbook": Object {
    "blockNumber": 1,
    "buy": Array [],
    "sell": Array [
      Object {
        "baseAmount": "3",
        "baseToken": undefined,
        "offerId": "1",
        "ownerId": "",
        "price": "1",
        "quoteAmount": "3",
        "quoteToken": undefined,
        "timestamp": 2019-10-01T00:00:00.000Z,
        "type": "buy",
      },
      Object {
        "baseAmount": "3",
        "baseToken": undefined,
        "offerId": "2",
        "ownerId": "",
        "price": "2",
        "quoteAmount": "6",
        "quoteToken": undefined,
        "timestamp": 2019-10-01T00:00:00.000Z,
        "type": "buy",
      },
      Object {
        "baseAmount": "3",
        "baseToken": undefined,
        "offerId": "3",
        "ownerId": "",
        "price": "4",
        "quoteAmount": "12",
        "quoteToken": undefined,
        "timestamp": 2019-10-01T00:00:00.000Z,
        "type": "buy",
      },
      Object {
        "baseAmount": "4",
        "baseToken": undefined,
        "offerId": "4",
        "ownerId": "",
        "price": "5",
        "quoteAmount": "20",
        "quoteToken": undefined,
        "timestamp": 2019-10-01T00:00:00.000Z,
        "type": "buy",
      },
    ],
  },
  "plan": undefined,
  "price": "1",
  "quoteToken": "DAI",
  "readyToProceed": false,
  "slippageLimit": "0.05",
  "submit": [Function],
  "view": "instantTradeForm",
}
`;

exports[`set price and amount 1`] = `
Object {
  "account": "0x1234",
  "amount": "1",
  "baseToken": "WETH",
  "collRatio": undefined,
  "dustLimitBase": "0.1",
  "dustLimitQuote": "0.1",
  "etherPriceUsd": "1",
  "gasEstimationStatus": "unset",
  "gasPrice": "0.01",
  "isSafeCollRatio": true,
  "kind": "buy",
  "leverage": undefined,
  "liquidationPrice": undefined,
  "maxAmount": "0",
  "maxTotal": "0",
  "messages": Array [],
  "mta": Object {
    "daiAllowance": false,
    "marginableAssets": Array [],
    "proxy": Object {
      "options": Object {
        "address": "",
      },
    },
    "state": "setup",
    "totalAssetValue": "0",
    "totalAvailableDebt": "0",
    "totalDebt": "0",
  },
  "orderbook": Object {
    "blockNumber": 1,
    "buy": Array [],
    "sell": Array [
      Object {
        "baseAmount": "3",
        "baseToken": undefined,
        "offerId": "1",
        "ownerId": "",
        "price": "1",
        "quoteAmount": "3",
        "quoteToken": undefined,
        "timestamp": 2019-10-01T00:00:00.000Z,
        "type": "buy",
      },
      Object {
        "baseAmount": "3",
        "baseToken": undefined,
        "offerId": "2",
        "ownerId": "",
        "price": "2",
        "quoteAmount": "6",
        "quoteToken": undefined,
        "timestamp": 2019-10-01T00:00:00.000Z,
        "type": "buy",
      },
      Object {
        "baseAmount": "3",
        "baseToken": undefined,
        "offerId": "3",
        "ownerId": "",
        "price": "4",
        "quoteAmount": "12",
        "quoteToken": undefined,
        "timestamp": 2019-10-01T00:00:00.000Z,
        "type": "buy",
      },
      Object {
        "baseAmount": "4",
        "baseToken": undefined,
        "offerId": "4",
        "ownerId": "",
        "price": "5",
        "quoteAmount": "20",
        "quoteToken": undefined,
        "timestamp": 2019-10-01T00:00:00.000Z,
        "type": "buy",
      },
    ],
  },
  "plan": undefined,
  "price": "1",
  "priceImpact": "0",
  "quoteToken": "DAI",
  "readyToProceed": false,
  "slippageLimit": "0.05",
  "submit": [Function],
  "total": "1",
  "view": "instantTradeForm",
}
`;
